crates/wasi-c2/src/sched/sync.rs

```diff
@@ -0,0 +1,131 @@
+    use yanix::poll::{PollFd, PollFlags};
+
+    #[derive(Default)]
+    pub struct SyncSched {}
+
+    impl WasiSched for SyncSched {
+        fn poll_oneoff<'a>(&self, poll: &'a Poll<'a>) -> Result<(), Error> {
+            if poll.is_empty() {
+                return Ok(());
+            }
+            let mut pollfds = Vec::new();
+            let timeout = poll.earliest_system_timer();
+            for s in poll.rw_subscriptions() {
+                match s {
+                    Subscription::Read(f) => {
+                        let raw_fd = wasi_file_raw_fd(f.file.deref()).ok_or(Error::Inval)?;
+                        pollfds.push(unsafe { PollFd::new(raw_fd, PollFlags::POLLIN) });
+                    }
+
+                    Subscription::Write(f) => {
+                        let raw_fd = wasi_file_raw_fd(f.file.deref()).ok_or(Error::Inval)?;
+                        pollfds.push(unsafe { PollFd::new(raw_fd, PollFlags::POLLOUT) });
+                    }
+                    Subscription::SystemTimer { .. } => unreachable!(),
+                }
+            }
+
+            let ready = loop {
+                let poll_timeout = if let Some(t) = timeout {
+                    let duration = t
+                        .deadline
+                        .duration_since(t.clock.now(t.precision))
+                        .unwrap_or(Duration::from_secs(0));
+                    (duration.as_millis() + 1) // XXX try always rounding up?
+                        .try_into()
+                        .map_err(|_| Error::Overflow)?
+                } else {
+                    libc::c_int::max_value()
+                };
+                tracing::debug!(
+                    poll_timeout = tracing::field::debug(poll_timeout),
+                    poll_fds = tracing::field::debug(&pollfds),
+                    "poll"
+                );
+                match yanix::poll::poll(&mut pollfds, poll_timeout) {
+                    Ok(ready) => break ready,
+                    Err(_) => {
+                        let last_err = std::io::Error::last_os_error();
+                        if last_err.raw_os_error().unwrap() == libc::EINTR {
+                            continue;
+                        } else {
+                            return Err(last_err.into());
+                        }
+                    }
+                }
+            };
+            if ready > 0 {
+                for (rwsub, pollfd) in poll.rw_subscriptions().zip(pollfds.into_iter()) {
+                    if let Some(revents) = pollfd.revents() {
+                        let (nbytes, rwsub) = match rwsub {
+                            Subscription::Read(sub) => {
+                                (1, sub)
+                                // XXX FIXME: query_nbytes in wasi-common/src/sys/poll.rs
+                                // uses metadata.len - tell to calculate for regular files,
+                                // ioctl(fd, FIONREAD) for large files
```

Comment:
system-interface has this in [`num_ready_bytes`](https://docs.rs/system-interface/0.4.0/system_interface/io/trait.ReadReady.html#tymethod.num_ready_bytes). One difference from how wasi-common does it is that system-interface will sometimes return 0 if the underlying device can't report how many bytes it can read. Here, since we just called `poll` and it told us that *something* was ready, we should do `let n = sub.num_ready_bytes()` and then `max(n, 1)`.

Reply:
Fixed now- thanks!